R21.0 js to ts

colman-hartinger opened the pull request:
Pull request for switching from JS to TS, I also added prettierjs for formatting so some of those changes are here as well. Some additional formatting should be done to have api.ts and cli.ts follow more typescript procedures so I'll write up some separate tickets for that.

bradymadams commented on src/api.ts:
```diff
@@ -67,12 +89,13 @@ class API {
   static get version() {
-    return (typeof THOR_VERSION === 'undefined' ? '21.0' : THOR_VERSION);
+    return '21.0';
+    // return (typeof THOR_VERSION === 'undefined' ? '21.0' : THOR_VERSION);
```
What would be the best way to get the version from package.json? Should we create an issue to do that?

colman-hartinger replied:
I think adding a .env file would be the best way to go about setting the THOR version

bradymadams replied:
We'd need to then update versions in the env and in package.json, right? Not terrible, but if we could read the version from package.json, I think that would be best. One place to increment the rev. I created an issue, so we can just ignore it for now.
https://github.com/tetonsim/thorjs/issues/28

bradymadams commented on src/api.ts:
```diff
@@ -81,15 +104,16 @@ class API {
-  _request(method, route, success, error, data) {
+  _request(method, route, success, error, data?) {
     const xhttp = new XMLHttpRequest();
+    var response;
```
Some of my comments on this review are just going to be general questions about TS since I'm unfamiliar. Does TS enforce scope, and that's why you had to move the declaration of `response` out of the try? Since it's only used (from what I can tell) inside the onreadystagechange callback function can we declare it in there?

bradymadams replied:
This still needs to be moved.

bradymadams commented on src/api.ts:
```diff
@@ -278,25 +324,32 @@ class API {
    * @param {API~error} error Callback function if refresh is unsuccessful.
    */
-  refreshToken(success, error) {
+  refreshToken(success: any, error: (arg0: string) => void) {
```
Can we rename `arg0` to something more meaningful? I assume it's the error message string. Same on line 346.

bradymadams commented on src/api.ts:
```diff
@@ -320,12 +373,7 @@ class API {
-      error,
-      {code: code},
-    );
+    this._request('POST', '/auth/verify', success, error, { code: code });
```
Is the linter making these changes? If so, what's the styling rule here? I noticed other places where the reverse change was made. Is it a certain line length?

colman-hartinger replied:
I was trying [prettier.js](https://prettier.io/) which is a code formatter rather than a linter. I liked the changes and thought it made the code a bit more readable so I added it to this PR. If you think the changes are strange we can get rid of it and just continue to use an "un-opinionated" formatter like eslint

bradymadams commented on src/api.ts:
```diff
@@ -712,7 +734,9 @@ class API {
-    this._request('DELETE', `/teams/${team}/member`, success, error, {email: email});
+    this._request('DELETE', `/teams/${team}/member`, success, error, {
+      email: email,
+    });
```
Line breaking in the middle of an argument without doing one argument per line looks strange to me and I've always avoided this formatting. I'm not totally against it, but is there a rule in the linter that's doing this that we could adjust?

colman-hartinger replied:
If you think this is strange we can just get rid of prettier, I was just trying it out but eslint should already catch any minor error so it's not necessary

bradymadams commented on src/thor.ts:
```diff
@@ -1,4 +1,6 @@
-const API = require('./api');
+export {};
```
TS question - what does this do?

colman-hartinger replied:
the empty export lets typesscript know that the file is a module

bradymadams commented on src/cli.ts:
```diff
@@ -190,18 +184,15 @@ function _multipleQuestions(questions, callback) {
+        rl.close();
+        callback(...answers);
+      }
+    });
```
These formatting changes are harder to read IMO. Linebreaking in the middle of a function argument looks like the function logic are arguments to the original function call (at least to my eyes).

colman-hartinger replied:
I was thinking that it was a bit easier to rid but Ill remove prettier and revert back to the old styling

bradymadams reviewed (changes requested):
It seems like this PR mostly consists of style changes. Since we just merged a big style change PR (#20) can you help me understand why these changes are all necessary? To be honest, most of the style changes on the linebreaking of function declarations and calls seem a little inconsistent and atypical to me.

colman-hartinger commented:
@bradymadams I reverted all the prettier changes so now this should just reflect the TS changes and the changes lint mad since I switched up some rules when switching over to TS

bradymadams commented on src/cli.ts:
```diff
@@ -194,7 +199,7 @@ function _multipleQuestions(questions, callback) {
       (answer) => {
         answers.push(answer);
-        if (questions.length > (i + 1)) {
+        if (questions.length > i + 1) {
```
The math part of my brain hates not using parentheses here. Is this a linter rule?

colman-hartinger replied:
I disabled the rule

bradymadams replied:
Thanks. I know that most programming languages still evaluate the RHS fully before comparing the two sides, but it still makes me stop and think about it everytime I see it. Especially in a language like JS which will very happily add `true + 1` and `false + 1`.

bradymadams reviewed (approved):
Looks good. Looks like there's a lot more work we can do to add types to all of the function calls, but I suppose we can do that later.
